Code doesn't compile in EndeavourOS (Linux)

I obtained the necessary libraries. The code doesn't compile on Arch. Maybe other Linuxes too.

The error is:

7z_C/CpuArch.c:78:12: error: type defaults to ‘int’ in declaration of ‘_b’ [-Wimplicit-int]


There's a problem on CpuArch.c in the 7zip code that causes an error...I noticed that it was testing to see if it should use ASM or not, and I thought "Let's just get rid of everything in this block of code so the file looks like:

#include "CpuArch.h"

#ifdef MY_CPU_X86_OR_AMD64


#endif

...and it compiles on my computer. But that error should probably be fixed. I'm not really skilled enough in this area or knowledgeable enough about this project to fix it though, so I just used the nuclear method on what seems to be unnecessary code for my (non-7zip using) purposes to get the emulator up and running for me...without 7zip support though, since that most definitely crashes the emulator because of the edit I mentioned.

--- a/snesreader/7z_C/CpuArch.c
+++ b/snesreader/7z_C/CpuArch.c
@@ -3,11 +3,11 @@
 
 #include "CpuArch.h"
 
+#ifdef MY_CPU_X86_OR_AMD64
+
 #ifndef _MSC_VER
 #include <cpuid.h>
 #endif
-
-#ifdef MY_CPU_X86_OR_AMD64
 
 #if (defined(_MSC_VER) && !defined(MY_CPU_AMD64)) || defined(__GNUC__)
 #define USE_ASM

--- a/snesreader/7z_C/CpuArch.c
+++ b/snesreader/7z_C/CpuArch.c
@@ -3,6 +3,10 @@
 
 #include "CpuArch.h"
 
+#ifndef _MSC_VER
+#include <cpuid.h>
+#endif
+
 #ifdef MY_CPU_X86_OR_AMD64
 
 #if (defined(_MSC_VER) && !defined(MY_CPU_AMD64)) || defined(__GNUC__)
@@ -50,7 +54,7 @@ static UInt32 CheckFlag(UInt32 flag)
 
 static void MyCPUID(UInt32 function, UInt32 *a, UInt32 *b, UInt32 *c, UInt32 *d)
 {
-  #ifdef USE_ASM
+  #if defined USE_ASM && defined _MSC_VER // non-windows ASM version seems to be broken
 
   #ifdef _MSC_VER
 
@@ -89,12 +93,25 @@ static void MyCPUID(UInt32 function, UInt32 *a, UInt32 *b, UInt32 *c, UInt32 *d)
   
   #else
 
+  #ifdef _MSC_VER
+
   int CPUInfo[4];
   __cpuid(CPUInfo, function);
   *a = CPUInfo[0];
   *b = CPUInfo[1];
   *c = CPUInfo[2];
   *d = CPUInfo[3];
+
+  #else
+
+  UInt32 a2, b2, c2, d2;
+  __cpuid(function, a2, b2, c2, d2);
+  *a = a2;
+  *b = b2;
+  *c = c2;
+  *d = d2;
+
+  #endif
 
   #endif
 }

